### PR TITLE
[D2M] Generalize virtual grid strategy for nonmatmul ops

### PR DESCRIPTION
**Pull request by vwellsTT**
### Ticket
Link to Github Issue

### Problem description
In other work beefing up GridSelection, I hit some errors where nonsquare grids on BH runners have very suboptimal allocations in some cases due to weird virtual grid spaghetti logic.

### What's changed
Simplify virtual grid selection s.t. we don't special case 2D to height/width sharded only--this greatly optimizes some edge cases.
Fix oob_val bug this exposes w.r.t. to folding undef and neginf in both directions

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/D2M/Transforms/GridSelection.cpp [hunk 1]
```diff
@@ -260,7 +227,8 @@ static void insertViewForTTNNDRAMTensor(Value operand,
 
   auto viewOutputLayout = ttcore::MetalLayoutAttr::get(
       builder.getContext(), baseMetalLayout.getLogicalShape(),
-      ttcore::MemorySpace::DeviceDRAM, ttcore::TensorMemoryLayout::Interleaved,
+      baseMetalLayout.getOobVal(), ttcore::MemorySpace::DeviceDRAM,
```
Comment (vmilosevic) [inline lib/Dialect/D2M/Transforms/GridSelection.cpp]: :warning: **clang\-diagnostic\-error** :warning:
no member named `` getOobVal `` in `` mlir::tt::ttcore::MetalLayoutAttr ``

File: lib/Dialect/D2M/Transforms/GridSelection.cpp [hunk 2]
```diff
@@ -463,29 +375,31 @@ static void applyCompositeViewUpdate(
       const int64_t tileHWIdx = concatDim - (logicalRank - 2);
       inputAlignments[concatDim] = (tileHWIdx >= 0) ? tileShape[tileHWIdx] : 1;
 
-      auto newLayout = ttcore::MetalLayoutAttr::get(
+      auto coordLayout = ttcore::MetalLayoutAttr::get(
           builder.getContext(), inputLayout.getLogicalShape(),
-          inputLayout.getMemorySpace(), inputLayout.getMemoryLayout(),
-          inputLayout.getCollapsedIntervals(), inputAlignments);
+          inputLayout.getOobVal(), inputLayout.getMemorySpace(),
```
Comment (vmilosevic) [inline lib/Dialect/D2M/Transforms/GridSelection.cpp]: :warning: **clang\-diagnostic\-error** :warning:
no member named `` getOobVal `` in `` mlir::tt::ttcore::MetalLayoutAttr ``

File: lib/Dialect/D2M/Transforms/GridSelection.cpp [hunk 3]
```diff
@@ -710,16 +682,10 @@ recreateGenericOp(d2m::GenericOp genericOp,
...
-  const EffectiveTargetGridRange &effectiveTargetGridRange =
-      result.effectiveTargetGridRange;
-  TT_assertv(effectiveTargetGridRange.shape.size() == 2u,
-             "Expected 2D effective target grid shape");
-  TT_assertv(effectiveTargetGridRange.offset.size() == 2u,
-             "Expected 2D effective target grid offset");
+  // effectiveTargetGrid is the generic's target grid (full device grid, or the
+  // range scoped by an enclosing d2m.spatial region). Per-operand grid
+  // decisions already carry their selected, physical, and alignment grids.
+  ArrayRef<int64_t> effectiveTargetGrid = result.effectiveTargetGrid;
```
Comment (vmilosevic) [inline lib/Dialect/D2M/Transforms/GridSelection.cpp]: :warning: **clang\-diagnostic\-error** :warning:
no member named `` effectiveTargetGrid `` in `` mlir::tt::d2m::GenericGridAnalysisResult ``

**Review by vmilosevic — COMMENTED** [review]
:warning: `Clang-Tidy` found issue(s) with the introduced code (1/1)
